Creates class WP_Super_Cache_Loader

stodorovic opened the pull request:
* Creates class `WP_Super_Cache_Loader`.

Replaces: #16.

schlessera commented:
@stodorovic Is this a WIP? The current code is not being executed right now and does nothing in its current form...

stodorovic commented:
@schlessera Thank you for looking into this!

Yes, I'm still working on this. I've added (in my tests) this line before [`WP_CLI::add_command`](https://make.wordpress.org/cli/handbook/internal-api/wp-cli-add-command/):
```php
new WP_Super_Cache_Loader;
```
It should be called before [action `muplugins_loaded`](https://codex.wordpress.org/Plugin_API/Action_Reference/muplugins_loaded). Is it good place or we should move it somewhere else? I'll add it when you confirm is it OK.

Many of these changes are already integrated in WPSC (master repo). I'm planning to check compatibility with  previous versions and add minimal requirement for version of WPSC. After this, I'll create new PR which will cleanup WP_Super_Cache_Command. (and add `__invoke` method instead _before_invoke_).

schlessera commented on src/WP_Super_Cache_Loader.php:
```diff
@@ -0,0 +1,110 @@
+	 *
+	 * @return void
+	 */
+	function maybe_load_config() {
```
`private function ...`

schlessera commented on src/WP_Super_Cache_Loader.php:
```diff
@@ -0,0 +1,110 @@
+/**
+ * Load WPSC files and config file if they aren't loaded.
+ */
+class WP_Super_Cache_Loader {
```
`final class ...`

schlessera commented on src/WP_Super_Cache_Loader.php:
```diff
@@ -0,0 +1,110 @@
+		}
+
+		if ( ! isset( $cache_enabled, $super_cache_enabled, $wp_cache_mod_rewrite, $wp_cache_debug_log )
+			&& ! @include( $wp_cache_config_file ) // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged
```
Error suppression should be avoided at all cost.

Why not just check whether the file is readable before including it?

stodorovic replied:
It's copied from WPSC. I agree with you that isn't good and I'm updating this code.

schlessera commented on src/WP_Super_Cache_Loader.php:
```diff
@@ -0,0 +1,110 @@
+			&& ! @include( $wp_cache_config_file ) // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged
+		) {
+			if ( defined( 'WPCACHEHOME' ) && ! emtpy( $wp_cache_config_file_sample ) ) {
+				@include( $wp_cache_config_file_sample ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged
```
Error suppression should be avoided at all cost.

Why not just check whether the file is readable before including it?

stodorovic replied:
I'll remove `@` for now and avoid file checking because this file should be installed together with other plugin files.

schlessera commented:
@stodorovic I've added a few comments.
This class should be triggered from within the command code, otherwise it will always be active on any WP-CLI command.

You can adjust the timing of the command itself through something like `@when before_wp_load`.

Let me know if you need further help.

schlessera commented:
@stodorovic Your current approach manually loads WordPress, which you should avoid. Let WP-CLI itself handle loading of WordPress instead.

I'm not entirely sure what the requirements are for WP Super Cache, but very generally, you should think about the timing needs, and then map these onto WP-CLI/WordPress hooks.

There's an overview of the available WP-CLI hooks here: https://make.wordpress.org/cli/handbook/internal-api/wp-cli-add-hook/#notes

stodorovic commented on src/WP_Super_Cache_Loader.php:
```diff
@@ -0,0 +1,169 @@
+		}
+
+		$this->register_hooks();
+		WP_CLI::get_runner()->load_wordpress();
```
I've tried this, but when `__invoke` is called, WP loading doesn't continue. Anyway, I'd like to make more clear solution and I've re-checked all possible issues in WPSC. I'm submitting new version without hooks (I've left the code if we change something in the future).
From other side, latest WPSC better handles globals and this loader only handle multisite installations.

stodorovic commented on src/WP_Super_Cache_Command.php:
```diff
@@ -5,16 +5,42 @@
+		$this->wpsc_version = $loader->get_wpsc_version();
+
+		if ( ! function_exists( 'wp_super_cache_enable' ) || empty( $this->wpsc_version ) ) {
+			WP_CLI::error( 'WP Super Cache needs to be enabled to use its WP-CLI commands.' );
```
It's hard to create good "closure" which handles all errors. Anyway, it's called just before invoke command and it's very little difference related to performance ( `before_invoke` checks whether function `wp_super_cache_enable` exists - it exists only when wp is loaded).
 
Probably it's better to use `__invoke` which check status of WP and shows error before run of specific command. We could improve it in next PR.

stodorovic commented:
@schlessera Sorry for delay. I've tested possible issues in WPSC and I've modified loader to avoid manually loading of WordPress.

Now, all commands runs when WordPress is loaded (as was before). I've added method `maybe_include_file` which includes PHP files if it needs (and populates globals). I've tested randomly various version of WPSC(between 1.5.2 and latest) and it seems that everything works correctly. Latest version of WPSC(1.6.4) contains https://github.com/Automattic/wp-super-cache/pull/592 which explicitly loads globals from wp-cache-config.php (and loader is needed only for multisite).

schlessera commented:
@stodorovic This is looking good already. I've had two nitpicks, but after these are resolved, I think we're ready to merge this iteration.

stodorovic commented:
I just applied your proposals and something which I discovered in meantime. I agree with you that's enough for this iteration. I'm preparing PRs which will improve `WP_Super_Cache_Command.php`.
